org_settings: Add feature to change bot role from UI.

somesh202 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
The first commit: 
- Adds a new column in the bot-list table in the organization settings to show the role a bot has been assigned.

The second commit: 

- Adds the `Role` field in the `Manage bot` modal, allowing an admin to change roles of a bot through the UI and fixing the 
  formatting and ordering of fields in the modal.

The third commit: 
- Adds the main feature to change the roles of a bot through the UI. 

Fixes: #21105

This PR is a follow-up to #21253 which was accidentally closed and the branches got deleted.
**Testing plan:** <!-- How have you tested? -->

**GIFs or screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->
![Screenshot 2022-02-25 171455](https://user-images.githubusercontent.com/77766761/155833545-78f1c102-7064-40fa-a4a0-fd35e1d57f77.png)
![Screenshot 2022-03-01 204419](https://user-images.githubusercontent.com/77766761/156195419-78a14eee-dcec-422a-9ae3-295b0dc56421.png)

<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

alya commented:
Replying to questions on the other PR:

> I am not sure too, but for various purposes we check the permission of its owner. @alya thoughts?

I don't think we should change the possible roles in this PR, or at least is should be a separate commit that we can decide to merge or not separately. @timabbott may have thoughts on whether bot should be allowed to be owners, but it's orthogonal to the other changes being made.

alya commented:
> @alya do we need the Bot type field here? I think it is not mentioned in the issue, so just wanted to confirm since commit message also does not explains adding this.

I don't think we need it, since it can't be changed... @timabbott what do you think?

sahil839 commented:
I tested it, sorting with role does not work correctly now.

somesh202 commented:
@sahil839  Yeah, I tried debugging it, but couldn't find a way out why the sorting is not happening as the `sort_role` function works correctly for the users_table.

sahil839 commented:
I guess this is because there is no role field in the object returned by `bot_info`.

somesh202 commented:
Ah, the role field is there as `info.user_role_text` in the `bot_info()` function as in the case of users.

sahil839 commented:
It is the text description of role field. Not the actual integer value of role. And in users list `human_info` is used only to render template and not as `get_item` function of list_widget.

somesh202 commented:
@sahil839 Thanks! I've updated the PR with the fix for sorting of role column in case of bots.
Also, I've removed the `bot_type` field for now as suggested by alya.

sahil839 commented:
LGTM.
@timabbott I think this can be reviewed by you. I have a couple of questions -
- What url should we use to change role? This PR currently uses `/json/users/{user_id}` url. If we want to use `/json/bots` url we need to add backend support too.
- Do we want to include the `role` field in the `realm_bots` object received from server?

timabbott commented:
The first two commit look good, but I think the two `do_settings_change` calls approach in the last commit doesn't make sense.

I think it probably makes the most sense to extend `/json/bots/{bot_id}` to allow changing a bot's role. We should do that as a documented API change in its own commit, and then rework the last commit to just use that endpoint.

I don't think we strictly need `role` in `realm_bots` in that I think it's in `realm_user` already ... and I kinda feel like the role of `realm_bots` is to have information that's additive (like the services details). But my opinion could change if there's something made awkward by the current model.

timabbott commented on static/js/settings_users.js:
```diff
@@ -682,10 +690,16 @@ function handle_bot_form($tbody, $status_field) {
+                url_for_role_change,
+                role_data,
+                $status_field,
+            );
```
As a request for future PRs, please call attention when you're doing something quite weird like this (i.e. making write requests to two endpoints) -- it shouldn't be the reviewer's responsibility to notice this.  A code comment would be helpful, but also a comment on the PR is great.

Generally things like this that are effectively workarounds tend to be things we want to find a better solution to before merging a change.

timabbott commented on static/templates/settings/admin_bot_form.hbs:
```diff
@@ -13,6 +13,14 @@
         </div>
+        <div class="input-group">
+            <label class="input-label" for="bot-role-select">{{t 'Role' }}
+                {{> ../help_link_widget link="/help/roles-and-permissions" }}
```
@alya @laurynmm should this link to the roles and permissions page in the API documentation instead? We can then potentially enhance that page with a new section about giving bot users roles.

alya replied:
What's the page? I'm not familiar with API documentation at all, and I don't see it in the left sidebar.

laurynmm replied:
@alya - We added that page recently and it looks like the general zulip docs haven't been updated for that yet. You can see it on the [CZO API docs](https://chat.zulip.org/api/roles-and-permissions).

I feel like adding information to the roles and permissions API page about bot users having roles could be useful, but in a context of how clients would use/need that information.

For this, I wonder if we want to update (and link to) the [about bots](https://chat.zulip.org/help/bots-and-integrations) page to be clearer about bot users roles, and link to the help center's roles and permissions doc from there?

alya replied:
As discussed on our call, that make sense to me.

timabbott replied:
@laurynmm can you make sure this gets captured in an issue or your TODO list? I'm planning to merge this PR.

laurynmm replied:
:+1:  I've got it on my to-do for now
